### ISSUE TITLE
New update system and launcher.

### ISSUE THREAD

**Issue by Gyrth** [moonwards1/Moonwards-Virtual-Moon#93]
**Step 1: Updater**
* Determine how updates are done in the future
	* A seperate application will serve as a launcher and updater
	* This can be written in Godot
* Write the launcher
	* It performs CRC checks for the game and is capable of launching it
	* It can download new files from a file server
	* To start with we will focus on redownloading files from scratch

**Comment by Gyrth** [comment]
Should this be done in a separate repo?

The files that need to be downloaded are an executable and the pck with all the project files. How do you do update checks with that? Unless you want the server application to create pck on the fly based on the clients list of assets. In which case how will the executable be updated once we move over to a newer version of Godot?

To which directory would the launcher/downloader write to? The user cache file or the install directory?

**Comment by TheDuriel** [comment]
Yes this should be a separate repo. @Moonwards Can you create a moonwards1/Moonwards-Launcher repo?

The binary and pck can automatically be exported at regular intervals. But that's a luxury. Until that point, manually exporting and uploading it somewhere we are allowed to download from will suffice. (Github can work as a start.)

The file server would store: The game files. Consisting of binaries and pck. And a text file containing the md5 hash of each file.

The launcher then does the following:

* Download the text file. 
* Compare the hashes in that file, with the hashes of the files it finds on disk. 
* If theres a mismatch, download and replace the file.

Downloading would be done to the User directory. As its the only safe place to write without admin privileges. On windows we could technically dl to the install dir. But thats not an option on mac and linux, so lets not make exceptions.

In  the future this same system can be used to check any combination of files, be they pck's or loose assets.

**Comment by Moonwards** [comment]
Done.
https://github.com/moonwards1/Moonwards-Launcher
Everyone on the current team should have received an invite. It is set up in the same PR oriented way with the same permissions.

**Comment by TheDuriel** [comment]
Launcher development will happen here: https://github.com/moonwards1/Moonwards-Launcher

A basic godot project has been set up. Same workflow as with this repo.

**Comment by Gyrth** [comment]
https://github.com/Gyrth/Moonwards-Launcher/tree/dev1
I've made a basic downloader in my fork. It gets a json file off my server and checks the md5 that are in there. If the file is missing or has a different md5 then it starts downloading them and writing to the user dir.

A few issues though. 
1. There needs to be a platform check for the launcher to download the right files. (Windows, MacOS or Linux)
2. The launch method ``OS.execute`` is different for each platform and a bit finicky.
3. Before Moonwards can be executed, the current working directory needs to be changed. Or else the .pck in the current directory will be used. (In this case the launcher) 
4. The downloaded file on Linux is not marked as executable even though I set that. And I'm not sure if Godot can change those permissions.

**Comment by Gyrth** [comment]
5. How are we going to update the launcher if there needs to be something fixed?
6. Putting the launcher inside the Moonwards project would actually be easier since you can just relaunch it once an update is pulled without changing the working directory.

**Comment by TheDuriel** [comment]
1. Seems simple enough. The OS singleton is aware of the platform.

2. Will need testing. Windows is thankfully easy enough. OS.execute() takes terminal arguments after all.

3. Same problem as 2. Just to clarify: the Launcher would also download the binary, and run that. So there is no risk of it running the launcher pck.

4. No clue. We should ask on the Godot board.

5. The Launcher can technically download itself. Much easier and more practical however is to simply check another hash file for the launcher version, and display a link for where to download the new one. We can do live updating later. (Which with the launcher being a much simpler project, will be much easier.)

6. Actually, can you clarify what your issue with the working dir is? Launcher and Game are sperate binaries. With separate names for binary and pcks. While they can coexist in the same directory, it is not feasible to download the game into the launchers directory, should it be placed in a location where it does not have the permission to do so. Which on Windows 7+ is _anywhere_ unless ran as Admin.

**Comment by Gyrth** [comment]
I wrote two files to the user directory using the launcher:
```xml
/home/gyrth/.local/share/godot/app_userdata/moonwards-launcher/MoonTown.pck
/home/gyrth/.local/share/godot/app_userdata/moonwards-launcher/MoonTown.x86_64
```

Now if I tried to run the ``MoonTown.x86_64`` using ``OS.execute()`` it would take the executable, but use the ``.pck`` from the launcher's ``res://`` directory.

To fix this I had to use ``cd`` first. Like this.
```swift
OS.execute('/bin/sh', ["-c", "cd " + OS.get_user_data_dir() + " && ./MoonTown.x86_64"], false, output)
```

**Comment by TheDuriel** [comment]
Ty OS.shell_open()

**Comment by Gyrth** [comment]
``var error = OS.shell_open(OS.get_user_data_dir() + "/MoonTown.x86_64")``
This doesn't seem to do anything. And the function returns 0 for OK.

**Comment by TheDuriel** [comment]
Just to make sure the path is fine try: `OS.get_user_data_dir().plus_file("MoonTown.x86_64")`

**Comment by TheDuriel** [comment]
If that won't work, I guess we'll have to find the correct OS.execute() for each platform.

**Comment by Gyrth** [comment]
I tried using ``OS.shell_open()`` for a directory and txt file. It seems to work for those. But not for an executable. :(

**Comment by Gyrth** [comment]
There seem to be two issues on Windows.

1. An error occurs while downloading new files, making the launcher try again and downloading in a loop.
2. Download speed is very low on my system, not sure if this affects others as well.

I've added the error code to the log system, so that should give us a good idea what's going on. Unfortunately I can't seem to reproduce this so if any of you could checkout dev3, press F1 to show the log UI and see what the error codes are that would be great. I've changed the md5 on the file server so it should loop updating for testing purposes. When it closes the current instance and starts a new one, then the update went correctly. If the same instance stays open and it keeps downloading the same file then you are running into the issue.

For the low download speed, may this could be the issue we're running into?
https://github.com/godotengine/godot/issues/32807
It is marked as fixed but is set to release with 3.2.

**Comment by TheDuriel** [comment]
We can keep an eye on the speed issue. We will most likely be migrating to 3.2 without issues, once the first few hotfixes, 3.2.X, releases are out.
